lib/document/stateTransition/documentTransition/AbstractDocumentTransition.js

```diff
@@ -22,17 +26,114 @@ class AbstractDocumentTransition {
     return this.dataContractId;
   }
 
+  /**
+   * Get data
+   *
+   * @return {Object}
+   */
+  getData() {
+    return {};
+  }
+
+  /**
+   * @abstract
+   *
+   * @return {Object}
+   */
+  getSystemProperties() {
+    throw new Error('Not implemented');
+  }
+
   /**
    * Get JSON representation
    *
-   * @returns { { $action: string, $dataContractId: string } }
+   * @return {Object}
    */
   toJSON() {
+    const data = JSON.parse(
+      JSON.stringify(this.getData()),
+    );
+
+    const encodedProperties = this.dataContract.getEncodedProperties(
+      this.getType(),
+    );
+
+    Object.keys(encodedProperties)
+      .forEach((propertyPath) => {
+        const property = encodedProperties[propertyPath];
+
+        if (property.contentEncoding === 'binary') {
+          const value = lodashGet(data, propertyPath);
+          if (value !== undefined) {
+            lodashSet(
+              data,
+              propertyPath,
+              Buffer.from(value)
+                .toString('utf-8'),
+            );
+          }
+        }
+      });
+
     return {
       $action: this.getAction(),
       $dataContractId: this.getDataContractId(),
+      ...this.getSystemProperties(),
+      ...data,
     };
   }
+
+  /**
+   * Get plain object representation
+   *
+   * @returns {Object}
+   */
+  toObject() {
+    return {
+      $action: this.getAction(),
+      $dataContractId: this.getDataContractId(),
+      ...this.getSystemProperties(),
+      ...this.getData(),
+    };
+  }
+
+  /**
+   * Create document transition from JSON
+   *
+   * @param {Class} documentTransitionClass
+   * @param {
+   *   RawDocumentCreateTransition | RawDocumentReplaceTransition | RawDocumentDeleteTransition
+   * } rawDocumentTransition
+   * @param {DataContract} dataContract
+   *
+   * @return {
+   *   RawDocumentCreateTransition | RawDocumentReplaceTransition | RawDocumentDeleteTransition
+   * }
+   */
+  static fromJson(documentTransitionClass, rawDocumentTransition, dataContract) {
```

Comment:
Let move it to specific transitions and create a protected method with common logic?